File: week1/Arrays/Level1/LongestSubArrayWithKSum.py
from typing import List
def LongestSubArrayWithKSum(nums: List[int], k: int) -> int:
    sum = res = 0
    map = {}
    for i in range(len(nums)):
        sum += nums[i]
        if sum == k:
            res = max(res,i+1)
        rem = sum - k
        if rem in map:
            l = i - map[rem]
            res = max(res,l)
        if rem not in map:
            map[rem] = i
    return res

nums = list(map(int,input().split()))
k = int(input())
print(LongestSubArrayWithKSum(nums,k))

File: week1/Arrays/Level1/LongestSubArrayWithKSum.js
const readline = require('readline')

const longestSubArrayWithKSum = (nums,k) => {
    let sum = 0;
    let res = 0;
    let myMap = new Map();
    for(let i=0;i<nums.length;i++){
        sum += nums[i];
        if(sum === k) res = Math.max(res,i+1);
        let rem = sum - k;
        if(myMap.has(rem)){
            let len = i - myMap.get(rem);
            res = Math.max(res,len);
        }
        if(!myMap.has(sum)) myMap.set(sum,i);
    }
    return res;
}

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

rl.question('',(userInput) => {
    const allInput = userInput.split(" ").map(Number);
    const k = allInput.pop();
    const nums = allInput;
    console.log(longestSubArrayWithKSum(nums,k));
    rl.close();
});